ramp4-pcar4/ramp4-pcar4
# Implement Feature Maptips (Non-Anchored)

Related issue: #192 

_No longer called hovertip. Maptip is the new name._

Maptips are implemented using `vue-tippy` and the method highlighted in discussion #521 (with a few changes) and R&D #522.

Some changes to note:
- In order for crosshair support, the `followCursor` option was not used on the `tippy` and instead the `offset` values were calculated
    - This allows for placement anywhere on the map given screen coordinates and is not dependent on the mouse
- `throttle` was used instead of `debounce` because it better serves the purpose for buffering changes to the `MaptipStore`
    - The [throttle-debounce](https://www.npmjs.com/package/throttle-debounce) package was added for this
- There were a few modifications to the `tooltip` model in the `MaptipStore` to better fit `RAMP` datatypes
- Anchored maptips were not implemented as part of this PR
    - More sophistication is required and since it is currently not implemented in RAMP2, it won't be implemented in R4MP until it is required
    - There are `getter/setter` methods in the `MapAPI` for all properties of a maptip, giving implementers more control over the maptip
 
Here is the [✨demo✨](http://ramp4-app.azureedge.net/demo/users/sharvenp/522/host/index.html)
To test:
- Loading some feature-based layers (`GraphicsLayer`, `FeatureLayer`, `CSVLayer`, ...)
- Mouse over them and see if the correct information is displayed
- Do the same using the crosshairs

**Note: The maptips do not support `MapImageLayer` since the `hitTest` does not register on these layers. This can be explored in a seperate issue. Currently RAMP2 does not support this either.**

~~**Note that `pnpm-lock` was re-created to integrate PR #530 changes into this PR - hence the large line insertion/deletion count**~~ This has been fixed.


---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/534)
